File: packages/cli/src/templates/cc-native/_cc-native/hooks/suggest-fresh-perspective.py
#!/usr/bin/env python3
"""
PostToolUse hook - suggests /fresh-perspective when user appears stuck.

Detection patterns:
1. Same error appearing 3+ times
2. Repeated edits to same file without resolution
3. Test failures after multiple fix attempts

Behavior: Suggests (doesn't force) running /fresh-perspective.
Non-blocking - always returns success.

Configuration (in _cc-native/plan-review.config.json):
  "stuckDetection": {
    "enabled": true,           // Set to false to disable entirely
    "errorThreshold": 3,       // Errors before suggesting
    "fileEditThreshold": 4,    // Edits to same file before suggesting
    "testFailureThreshold": 3, // Test failures before suggesting
    "cooldown": 10,            // Tool calls between suggestions
    "maxSuggestions": 3        // Max suggestions per session
  }
"""

import sys

# TODO: Remove this early exit when TypeScript implementation is complete.
# The Python shared library (_shared/lib/) was deleted during the TS migration,
# so all imports from base.hook_utils / base.logger / lib.context fail.
sys.exit(0)

import json
import os
import re
from pathlib import Path
from typing import Any, Dict

# Add lib directories to path for imports
_hook_dir = Path(__file__).resolve().parent
_lib_dir = _hook_dir.parent / "lib"
_shared = _hook_dir.parent.parent / "_shared"
_shared_lib = _shared / "lib"
sys.path.insert(0, str(_lib_dir))
sys.path.insert(0, str(_shared))
sys.path.insert(0, str(_shared_lib))

from base.hook_utils import emit_context
from base.logger import log_debug, log_info, log_warn, log_error
from lib.context.context_store import get_context_by_session_id, save_state


# ---------------------------
# Configuration (defaults, overridden by config.json)
# ---------------------------

DEFAULT_CONFIG = {
    "enabled": True,
    "errorThreshold": 3,
    "fileEditThreshold": 4,
    "testFailureThreshold": 3,
    "cooldown": 10,
    "maxSuggestions": 3,
}


def _int_or_default(value: Any, default: int) -> int:
    """Coerce value to int, return default if not possible.

    Handles string numbers, floats, and invalid types gracefully.
    """
    if isinstance(value, int):
        return value
    if isinstance(value, float):
        return int(value)
    if isinstance(value, str):
        try:
            return int(value)
        except ValueError:
            return default
    return default


def load_config(project_dir: Path) -> Dict[str, Any]:
    """Load stuckDetection config from _cc-native/plan-review.config.json."""
    config_path = project_dir / "_cc-native" / "plan-review.config.json"
    if not config_path.exists():
        return DEFAULT_CONFIG.copy()
    try:
        full_config = json.loads(config_path.read_text(encoding="utf-8"))
        section = full_config.get("stuckDetection", {})
        return {**DEFAULT_CONFIG, **section}
    except Exception as e:
        log_warn("suggest-fresh-perspective", f"Failed to load config: {e}")
        return DEFAULT_CONFIG.copy()


def get_project_dir(payload: Dict[str, Any]) -> Path:
    """Get project directory from payload or environment."""
    p = os.environ.get("CLAUDE_PROJECT_DIR") or payload.get("cwd") or os.getcwd()
    return Path(p)


# ---------------------------
# Compiled patterns (performance optimization)
# ---------------------------

# Single combined pattern for error detection (case-insensitive)
_ERROR_PATTERN = re.compile(r"(error:|failed|exception)", re.IGNORECASE)

# Combined pattern for test failures
_TEST_FAILURE_PATTERN = re.compile(
    r"(\d+\s+failed|FAIL\s|✗|AssertionError|test.*failed|npm\s+ERR!.*test)",
    re.IGNORECASE,
)

# Pattern for normalizing error messages (line numbers)
_LINE_NUMBER_PATTERN = re.compile(r":\d+")
_MULTI_DIGIT_PATTERN = re.compile(r"\d{2,}")
_PATH_PATTERN = re.compile(r"[/\\][^\s/\\]+[/\\]")


# ---------------------------
# State management (state.json integration)
# ---------------------------


def _get_cc_native_state(session_id: str, project_root: Path) -> Dict[str, Any]:
    """Get cc_native state from context state.json.

    Returns the cc_native dict or empty dict if not found.
    """
    try:
        state = get_context_by_session_id(session_id, project_root)
        if state and hasattr(state, "cc_native"):
            return state.cc_native
    except Exception:
        pass
    return {}


def _save_cc_native_state(
    session_id: str, project_root: Path, cc_native_data: Dict[str, Any]
) -> bool:
    """Save cc_native state to context state.json.

    Returns True on success, False on failure.
    """
    try:
        state = get_context_by_session_id(session_id, project_root)
        if state:
            state.cc_native = cc_native_data
            save_state(state, project_root)
            return True
    except Exception as e:
        log_warn("suggest-fresh-perspective", f"Failed to save cc_native state: {e}")
    return False


def load_stuck_state(session_id: str, project_root: Path) -> Dict[str, Any]:
    """Load stuck detection state for this session from state.json."""
    cc_native = _get_cc_native_state(session_id, project_root)
    stuck_state = cc_native.get("stuck_detection", {})
    return {
        "error_hashes": stuck_state.get("error_hashes", {}),
        "file_edits": stuck_state.get("file_edits", {}),
        "test_failures": stuck_state.get("test_failures", 0),
        "tool_calls_since_suggestion": stuck_state.get(
            "tool_calls_since_suggestion", 0
        ),
        "suggestion_count": stuck_state.get("suggestion_count", 0),
    }


def save_stuck_state(
    session_id: str, project_root: Path, stuck_state: Dict[str, Any]
) -> None:
    """Save stuck detection state for this session to state.json."""
    try:
        cc_native = _get_cc_native_state(session_id, project_root) or {}
        cc_native["stuck_detection"] = stuck_state
        if not _save_cc_native_state(session_id, project_root, cc_native):
            log_warn(
                "suggest-fresh-perspective",
                f"Failed to save stuck detection state for session {session_id}",
            )
    except Exception as e:
        log_warn(
            "suggest-fresh-perspective", f"Failed to save stuck detection state: {e}"
        )


# ---------------------------
# Detection logic
# ---------------------------


def hash_error(error_text: str) -> str:
    """Create a simple hash of an error message for deduplication.

    Normalizes by removing line numbers and multi-digit numbers,
    but preserves enough context to distinguish different errors.
    """
    # Normalize: remove line numbers, preserve error type
    normalized = _LINE_NUMBER_PATTERN.sub(":N", error_text)
    normalized = _MULTI_DIGIT_PATTERN.sub("N", normalized)
    # Simplify paths but keep some structure
    normalized = _PATH_PATTERN.sub(".../", normalized)
    # Take first 100 chars after normalization
    return normalized[:100]


def detect_repeated_error(
    state: Dict[str, Any], tool_result: str, threshold: int
) -> bool:
    """Check if we're seeing the same error repeatedly.

    Returns True if threshold reached, always updates state.
    """
    if not tool_result:
        return False

    if _ERROR_PATTERN.search(tool_result):
        error_hash = hash_error(tool_result)
        state["error_hashes"][error_hash] = state["error_hashes"].get(error_hash, 0) + 1
        return state["error_hashes"][error_hash] >= threshold

    return False


def detect_repeated_file_edits(
    state: Dict[str, Any], tool_name: str, tool_input: Dict[str, Any], threshold: int
) -> bool:
    """Check if we're editing the same file repeatedly.

    Returns True if threshold reached, always updates state.
    """
    if tool_name != "Edit":
        return False

    # Validate tool_input is a dict
    if not isinstance(tool_input, dict):
        return False

    file_path = tool_input.get("file_path", "")
    if not file_path:
        return False

    state["file_edits"][file_path] = state["file_edits"].get(file_path, 0) + 1
    return state["file_edits"][file_path] >= threshold


def detect_test_failures(
    state: Dict[str, Any], tool_name: str, tool_result: str, threshold: int
) -> bool:
    """Check for repeated test failures.

    Returns True if threshold reached, always updates state.
    """
    if tool_name != "Bash":
        return False

    if _TEST_FAILURE_PATTERN.search(tool_result):
        state["test_failures"] = state.get("test_failures", 0) + 1
        return state["test_failures"] >= threshold

    return False


# ---------------------------
# Main hook logic
# ---------------------------


def should_suggest(state: Dict[str, Any], cooldown: int) -> bool:
    """Check if we're past the cooldown period."""
    return state.get("tool_calls_since_suggestion", 0) >= cooldown


def create_suggestion() -> None:
    """Emit the suggestion via hook utility."""
    emit_context(
        "\n---\n"
        "**Stuck?** You've been working on similar issues for a while. "
        "Consider running `/fresh-perspective` to get an unbiased view of the problem "
        "without code context anchoring your thinking.\n"
        "---\n"
    )


def main() -> int:
    # === FAST PATH: Cheap checks first, no I/O ===

    try:
        payload = json.load(sys.stdin)
    except json.JSONDecodeError:
        return 0  # Fail-safe

    # 1. Check hook_event_name (cheap dict lookup)
    if payload.get("hook_event_name") != "PostToolUse":
        return 0

    # 2. Check session_id exists (cheap dict lookup)
    session_id = payload.get("session_id")
    if not session_id:
        return 0

    # 3. Check tool_name is relevant (cheap dict lookup)
    # We only care about Edit and Bash - skip everything else
    tool_name = payload.get("tool_name", "")
    if tool_name not in ("Edit", "Bash"):
        return 0

    # === SLOW PATH: Only reached for Edit/Bash tools ===

    # Load configuration (file I/O)
    project_dir = get_project_dir(payload)
    config = load_config(project_dir)

    # Check if feature is disabled
    if not config.get("enabled", True):
        return 0

    tool_input = payload.get("tool_input", {})
    tool_result = payload.get("tool_result", {})

    # Validate tool_input type
    if not isinstance(tool_input, dict):
        tool_input = {}

    # Extract result text
    result_text = ""
    if isinstance(tool_result, dict):
        result_text = str(
            tool_result.get("output", "") or tool_result.get("content", "")
        )
    elif isinstance(tool_result, str):
        result_text = tool_result

    # Load state from state.json
    state = load_stuck_state(session_id, project_dir)

    # Increment tool call counter
    state["tool_calls_since_suggestion"] = (
        state.get("tool_calls_since_suggestion", 0) + 1
    )

    # Get thresholds from config (with type coercion for safety)
    error_threshold = _int_or_default(config.get("errorThreshold"), 3)
    file_edit_threshold = _int_or_default(config.get("fileEditThreshold"), 4)
    test_failure_threshold = _int_or_default(config.get("testFailureThreshold"), 3)
    cooldown = _int_or_default(config.get("cooldown"), 10)
    max_suggestions = _int_or_default(config.get("maxSuggestions"), 3)

    # Run ALL detections (don't short-circuit - each updates state)
    error_detected = detect_repeated_error(state, result_text, error_threshold)
    file_edit_detected = detect_repeated_file_edits(
        state, tool_name, tool_input, file_edit_threshold
    )
    test_failure_detected = detect_test_failures(
        state, tool_name, result_text, test_failure_threshold
    )

    # Save state AFTER all detections have run
    save_stuck_state(session_id, project_dir, state)

    # Check if any detection triggered
    is_stuck = error_detected or file_edit_detected or test_failure_detected

    if is_stuck:
        if error_detected:
            log_info("suggest-fresh-perspective", "Detected repeated error pattern")
        if file_edit_detected:
            log_info("suggest-fresh-perspective", "Detected repeated file edits")
        if test_failure_detected:
            log_info("suggest-fresh-perspective", "Detected repeated test failures")

    # Only suggest if stuck AND past cooldown
    if is_stuck and should_suggest(state, cooldown):
        # Reset cooldown
        state["tool_calls_since_suggestion"] = 0
        state["suggestion_count"] = state.get("suggestion_count", 0) + 1
        save_stuck_state(session_id, project_dir, state)

        # Only suggest up to maxSuggestions times per session
        if state["suggestion_count"] <= max_suggestions:
            log_info(
                "suggest-fresh-perspective",
                f"Suggesting fresh perspective (suggestion #{state['suggestion_count']})",
            )
            create_suggestion()

    return 0


if __name__ == "__main__":
    from base.hook_utils import run_hook

    run_hook(main, "suggest_fresh_perspective")

File: packages/cli/src/templates/cc-native/_cc-native/hooks/suggest-fresh-perspective.ts
#!/usr/bin/env bun
/**
 * PostToolUse hook — suggests /fresh-perspective when user appears stuck.
 *
 * Detection patterns:
 * 1. Same error appearing 3+ times
 * 2. Repeated edits to same file without resolution
 * 3. Test failures after multiple fix attempts
 *
 * Behavior: Suggests (doesn't force) running /fresh-perspective.
 * Non-blocking — always returns success.
 *
 * Configuration (in _cc-native/plan-review.config.json):
 *   "stuckDetection": {
 *     "enabled": true,           // Set to false to disable entirely
 *     "errorThreshold": 3,       // Errors before suggesting
 *     "fileEditThreshold": 4,    // Edits to same file before suggesting
 *     "testFailureThreshold": 3, // Test failures before suggesting
 *     "cooldown": 10,            // Tool calls between suggestions
 *     "maxSuggestions": 3        // Max suggestions per session
 *   }
 */

import * as fs from "node:fs";
import * as path from "node:path";
import {
  loadHookInput,
  runHook,
  logDebug,
  logInfo,
  logWarn,
  emitContext,
} from "../../_shared/lib-ts/base/hook-utils.js";
import { getProjectRoot } from "../../_shared/lib-ts/base/constants.js";
import { getContextBySessionId, saveState } from "../../_shared/lib-ts/context/context-store.js";
import type { ContextState } from "../../_shared/lib-ts/types.js";

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

interface StuckConfig {
  enabled: boolean;
  errorThreshold: number;
  fileEditThreshold: number;
  testFailureThreshold: number;
  cooldown: number;
  maxSuggestions: number;
}

const DEFAULT_CONFIG: StuckConfig = {
  enabled: true,
  errorThreshold: 3,
  fileEditThreshold: 4,
  testFailureThreshold: 3,
  cooldown: 10,
  maxSuggestions: 3,
};

function intOrDefault(value: unknown, fallback: number): number {
  if (typeof value === "number") return Math.floor(value);
  if (typeof value === "string") {
    const n = parseInt(value, 10);
    return isNaN(n) ? fallback : n;
  }
  return fallback;
}

function loadStuckConfig(projectDir: string): StuckConfig {
  const configPath = path.join(projectDir, "_cc-native", "plan-review.config.json");
  if (!fs.existsSync(configPath)) return { ...DEFAULT_CONFIG };
  try {
    const raw = fs.readFileSync(configPath, "utf-8");
    const full = JSON.parse(raw) as Record<string, unknown>;
    const section = (full.stuckDetection ?? {}) as Partial<StuckConfig>;
    return { ...DEFAULT_CONFIG, ...section };
  } catch (e) {
    logWarn("suggest-fresh-perspective", `Failed to load config: ${e}`);
    return { ...DEFAULT_CONFIG };
  }
}

// ---------------------------------------------------------------------------
// Compiled Patterns
// ---------------------------------------------------------------------------

const ERROR_PATTERN = /(?:error:|failed|exception)/i;
const TEST_FAILURE_PATTERN = /(?:\d+\s+failed|FAIL\s|✗|AssertionError|test.*failed|npm\s+ERR!.*test)/i;
const LINE_NUMBER_PATTERN = /:\d+/g;
const MULTI_DIGIT_PATTERN = /\d{2,}/g;
const PATH_PATTERN = /[/\\][^\s/\\]+[/\\]/g;

// ---------------------------------------------------------------------------
// State Management
// ---------------------------------------------------------------------------

interface StuckState {
  error_hashes: Record<string, number>;
  file_edits: Record<string, number>;
  test_failures: number;
  tool_calls_since_suggestion: number;
  suggestion_count: number;
}

/** Extended ContextState with cc_native method-specific data */
type CcNativeContextState = ContextState & { cc_native?: Record<string, unknown> };

function getCcNativeState(sessionId: string, projectRoot: string): Record<string, unknown> {
  try {
    const state = getContextBySessionId(sessionId, projectRoot) as CcNativeContextState | null;
    if (state?.cc_native && typeof state.cc_native === "object") {
      return state.cc_native;
    }
  } catch { /* fail-safe */ }
  return {};
}

function saveCcNativeState(
  sessionId: string,
  projectRoot: string,
  ccNativeData: Record<string, unknown>,
): boolean {
  try {
    const state = getContextBySessionId(sessionId, projectRoot) as CcNativeContextState | null;
    if (state) {
      state.cc_native = ccNativeData;
      saveState(state.id, state, projectRoot);
      return true;
    }
  } catch (e) {
    logWarn("suggest-fresh-perspective", `Failed to save cc_native state: ${e}`);
  }
  return false;
}

function loadStuckState(sessionId: string, projectRoot: string): StuckState {
  const ccNative = getCcNativeState(sessionId, projectRoot);
  const stuck = (ccNative.stuck_detection ?? {}) as Record<string, unknown>;
  return {
    error_hashes: (stuck.error_hashes && typeof stuck.error_hashes === "object" ? stuck.error_hashes : {}) as Record<string, number>,
    file_edits: (stuck.file_edits && typeof stuck.file_edits === "object" ? stuck.file_edits : {}) as Record<string, number>,
    test_failures: typeof stuck.test_failures === "number" ? stuck.test_failures : 0,
    tool_calls_since_suggestion: typeof stuck.tool_calls_since_suggestion === "number" ? stuck.tool_calls_since_suggestion : 0,
    suggestion_count: typeof stuck.suggestion_count === "number" ? stuck.suggestion_count : 0,
  };
}

function saveStuckState(
  sessionId: string,
  projectRoot: string,
  stuckState: StuckState,
): void {
  try {
    const ccNative = getCcNativeState(sessionId, projectRoot) || {};
    ccNative.stuck_detection = stuckState;
    if (!saveCcNativeState(sessionId, projectRoot, ccNative)) {
      logWarn("suggest-fresh-perspective", `Failed to save stuck detection state for session ${sessionId}`);
    }
  } catch (e) {
    logWarn("suggest-fresh-perspective", `Failed to save stuck detection state: ${e}`);
  }
}

// ---------------------------------------------------------------------------
// Detection Logic
// ---------------------------------------------------------------------------

function hashError(errorText: string): string {
  let normalized = errorText.replace(LINE_NUMBER_PATTERN, ":N");
  normalized = normalized.replace(MULTI_DIGIT_PATTERN, "N");
  normalized = normalized.replace(PATH_PATTERN, ".../");
  return normalized.slice(0, 100);
}

function detectRepeatedError(state: StuckState, toolResult: string, threshold: number): boolean {
  if (!toolResult) return false;
  if (ERROR_PATTERN.test(toolResult)) {
    const errorHash = hashError(toolResult);
    state.error_hashes[errorHash] = (state.error_hashes[errorHash] ?? 0) + 1;
    return state.error_hashes[errorHash]! >= threshold;
  }
  return false;
}

function detectRepeatedFileEdits(
  state: StuckState,
  toolName: string,
  toolInput: Record<string, unknown>,
  threshold: number,
): boolean {
  if (toolName !== "Edit") return false;
  if (!toolInput || typeof toolInput !== "object") return false;
  const filePath = String(toolInput.file_path ?? "");
  if (!filePath) return false;
  state.file_edits[filePath] = (state.file_edits[filePath] ?? 0) + 1;
  return state.file_edits[filePath]! >= threshold;
}

function detectTestFailures(
  state: StuckState,
  toolName: string,
  toolResult: string,
  threshold: number,
): boolean {
  if (toolName !== "Bash") return false;
  if (TEST_FAILURE_PATTERN.test(toolResult)) {
    state.test_failures += 1;
    return state.test_failures >= threshold;
  }
  return false;
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

function main(): void {
  const payload = loadHookInput();
  if (!payload) return;

  // Fast path: cheap checks first
  if (payload.hook_event_name !== "PostToolUse") return;

  const sessionId = payload.session_id;
  if (!sessionId) return;

  const toolName = payload.tool_name ?? "";
  if (toolName !== "Edit" && toolName !== "Bash") return;

  // Slow path: only for Edit/Bash
  const projectDir = getProjectRoot(payload.cwd);
  const config = loadStuckConfig(projectDir);

  if (!config.enabled) return;

  const toolInput = (payload.tool_input && typeof payload.tool_input === "object")
    ? payload.tool_input as Record<string, unknown>
    : {};

  const toolResult = payload.tool_result;
  let resultText = "";
  if (toolResult && typeof toolResult === "object") {
    const tr = toolResult as Record<string, unknown>;
    resultText = String(tr.output ?? tr.content ?? "");
  } else if (typeof toolResult === "string") {
    resultText = toolResult;
  }

  // Load state
  const state = loadStuckState(String(sessionId), projectDir);
  state.tool_calls_since_suggestion += 1;

  // Get thresholds
  const errorThreshold = intOrDefault(config.errorThreshold, 3);
  const fileEditThreshold = intOrDefault(config.fileEditThreshold, 4);
  const testFailureThreshold = intOrDefault(config.testFailureThreshold, 3);
  const cooldown = intOrDefault(config.cooldown, 10);
  const maxSuggestions = intOrDefault(config.maxSuggestions, 3);

  // Run ALL detections (don't short-circuit — each updates state)
  const errorDetected = detectRepeatedError(state, resultText, errorThreshold);
  const fileEditDetected = detectRepeatedFileEdits(state, toolName, toolInput, fileEditThreshold);
  const testFailureDetected = detectTestFailures(state, toolName, resultText, testFailureThreshold);

  // Save state AFTER all detections
  saveStuckState(String(sessionId), projectDir, state);

  const isStuck = errorDetected || fileEditDetected || testFailureDetected;

  if (isStuck) {
    if (errorDetected) logInfo("suggest-fresh-perspective", "Detected repeated error pattern");
    if (fileEditDetected) logInfo("suggest-fresh-perspective", "Detected repeated file edits");
    if (testFailureDetected) logInfo("suggest-fresh-perspective", "Detected repeated test failures");
  }

  // Only suggest if stuck AND past cooldown
  if (isStuck && state.tool_calls_since_suggestion >= cooldown) {
    state.tool_calls_since_suggestion = 0;
    state.suggestion_count += 1;
    saveStuckState(String(sessionId), projectDir, state);

    if (state.suggestion_count <= maxSuggestions) {
      logInfo("suggest-fresh-perspective", `Suggesting fresh perspective (suggestion #${state.suggestion_count})`);
      emitContext(
        "\n---\n" +
        "**Stuck?** You've been working on similar issues for a while. " +
        "Consider running `/fresh-perspective` to get an unbiased view of the problem " +
        "without code context anchoring your thinking.\n" +
        "---\n",
      );
    }
  }
}

runHook(main, "suggest_fresh_perspective");
